src/core/reactor_backend.cc

```diff
@@ -1463,29 +1489,142 @@ class reactor_backend_uring final : public reactor_backend {
+        return readable_or_writeable(listenfd).then([&listenfd, this] {
+            listenfd.maybe_no_more_recv();
+            return do_with(socket_address{}, [&listenfd, this] (auto& sa) {
+                auto req = internal::io_request::make_accept(listenfd.fd.get(), &sa.as_posix_sockaddr(), &sa.addr_length, SOCK_NONBLOCK | SOCK_CLOEXEC);
+                return queue_request(std::move(req)).then([&listenfd, &sa] (size_t result) {
+                    // Speculate that there is another connection on this listening socket
+                    // in case we are facing a connection storm.
+                    listenfd.speculate_epoll(EPOLLIN);
+                    pollable_fd pfd(file_desc::from_fd(result), pollable_fd::speculation(EPOLLOUT));
+                    return make_ready_future<std::tuple<pollable_fd, socket_address>>(std::move(pfd), std::move(sa));
+                });
+            });
+        }).handle_exception_type([&listenfd, this] (const std::system_error &e) {
+            if (e.code() == std::errc::resource_unavailable_try_again) {
+                // We speculated that we will have an another connection, but got
+                // a false alarm. Try again.
+                return accept(listenfd);
+            }
+            if (e.code() == std::errc::invalid_argument) {
+                // The chances are that we shutting down the connection.
+                listenfd.maybe_no_more_recv();
+            }
+            throw;
+        });
     }
     virtual future<> connect(pollable_fd_state& fd, socket_address& sa) override {
-        return _r.do_connect(fd, sa);
+        return do_with(sa, [&fd, this] (auto& sa) {
+            auto req = internal::io_request::make_connect(fd.fd.get(), &sa.as_posix_sockaddr(), sa.length());
+            return queue_request(std::move(req)).handle_exception_type([] (const std::system_error &e) {
+                if (e.code() == std::errc::operation_in_progress) {
+                    return make_ready_future<size_t>(0);
+                }
+                throw;
+            });
+        }).then([this, &fd] (size_t) {
+            return writeable(fd);
+        }).then([&fd] {
+            auto err = fd.fd.getsockopt<int>(SOL_SOCKET, SO_ERROR);
+            if (err != 0) {
+                throw std::system_error(err, std::system_category());
+            }
+            return make_ready_future();
+        });
     }
     virtual void shutdown(pollable_fd_state& fd, int how) override {
         fd.fd.shutdown(how);
     }
     virtual future<size_t> read_some(pollable_fd_state& fd, void* buffer, size_t len) override {
-        return _r.do_read_some(fd, buffer, len);
+        return readable(fd).then([this, &fd, buffer, len] () mutable {
+            return queue_request(internal::io_request::make_read(fd.fd.get(), -1, buffer, len, false));
+        }).then([&fd, len] (size_t result) {
+            if (result == len) {
+                fd.speculate_epoll(EPOLLIN);
```

Comment:
still need to find a way to handle `-EAGAIN` properly if we happen to run out of the read buffer and the last speculation fails. currently, `io_completion::complete_with()` translates `-EAGAIN` in `sqe->res` to a `std::system_error(EAGAIN, std::system_category())`.

Reply:
We just have to loop through readable() again.

However, if reading without SOCK_NONBLOCK, it will never return EAGAIN.